src/rgw/rgw_bucket.cc

```diff
@@ -818,107 +933,182 @@ int RGWBucket::init(RGWRados *storage, RGWBucketAdminOpState& op_state)
   return 0;
 }
 
-int RGWBucket::link(RGWBucketAdminOpState& op_state, std::string *err_msg)
+int RGWBucket::link(RGWBucketAdminOpState& op_state,
+	map<string, bufferlist>& attrs, std::string *err_msg)
 {
   if (!op_state.is_user_op()) {
     set_err_msg(err_msg, "empty user id");
     return -EINVAL;
   }
 
   string bucket_id = op_state.get_bucket_id();
-  if (bucket_id.empty()) {
-    set_err_msg(err_msg, "empty bucket instance id");
-    return -EINVAL;
-  }
 
   std::string display_name = op_state.get_user_display_name();
-  rgw_bucket bucket = op_state.get_bucket();
+  rgw_bucket& bucket = op_state.get_bucket();
+  if (!bucket_id.empty() && bucket_id != bucket.bucket_id) {
+    set_err_msg(err_msg,
+	"specified bucket id does not match " + bucket.bucket_id);
```

Comment:
@cbodley, I have something written. Could you tell me how I can test having a bucket with a non-matching id?

Reply:
we now delete the old bucket instance after reshard succeeds (this wasn't always the case), so we can't really use that to test this. but since multisite doesn't remove bucket instances for deleted buckets, we can test with that. for example, on the master zone:
* create a testbucket, get its bucket id with `radosgw-admin bucket stats --bucket testbucket`
* delete the testbucket
* recreate the testbucket (it gets a new bucket id)
* unlink the bucket with `radosgw-admin bucket unlink --uid <uid> --bucket testbucket`
* relink the bucket with `radosgw-admin bucket link --uid <uid> --bucket testbucket --bucket-id <initial bucket id>`

trying this myself, i see that this functionality is broken already; the `bucket link` command doesn't change the bucket "id", and leaves it pointing at the new bucket instance. maybe we should defer this feature to separate pr?

Reply:
ok, I'll open a tracker to follow it up. Thanks.